source/JABHandler.py

```diff
@@ -630,11 +630,23 @@ def event_gainFocus(vmID,accContext,hwnd):
 	eventHandler.queueEvent("gainFocus",obj)
 
 @AccessBridge_PropertyActiveDescendentChangeFP
-def internal_event_activeDescendantChange(vmID, event,source,oldDescendant,newDescendant):
-	hwnd=getWindowHandleFromAccContext(vmID,source)
-	internalQueueFunction(event_gainFocus,vmID,newDescendant,hwnd)
-	for accContext in [event,oldDescendant]:
-		bridgeDll.releaseJavaObject(vmID,accContext)
+def internal_event_activeDescendantChange(vmID, event, source, oldDescendant, newDescendant):
+	hwnd = getWindowHandleFromAccContext(vmID, source)
+	sourceContext = JABContext(hwnd=hwnd, vmID=vmID, accContext=source)
+	if internal_hasFocus(sourceContext):
+		internalQueueFunction(event_gainFocus, vmID, newDescendant, hwnd)
+	for accContext in [event, oldDescendant]:
+		bridgeDll.releaseJavaObject(vmID, accContext)
+
+
+def internal_hasFocus(sourceContext):
+	focus = api.getFocusObject()
+	while isinstance(focus, NVDAObjects.JAB.JAB):
+		if focus.jabContext == sourceContext:
+			return True
+		focus = focus.parent

```

Comment:
Instead of traversing ancestors like this, please consider using api.getFocusAncestors to get the ancestors.
I'm also not sure whether traversing ancestry is really necessary. What happens if you only check the focus object?

Reply:
I used api.getFocusAncestors.
Traversing ancestors is necessary because source, old descendant, and new descendant are all not equal to focused object in case of changing selection in focused tree by arrow keys.